### ISSUE TITLE
Cached projects are only restored the "target" folder but not the ~/.m2/repository folder

### ISSUE THREAD

**Issue by SchwarzAmihay** [tinesoft/nxrocks#215]
### Plugin Name

@nxrocks/nx-spring-boot

### Nx Report

 >  NX   Report complete - copy this into the issue template

   Node   : 18.17.0
   OS     : darwin-arm64
   pnpm   : 8.15.3
   
   nx                 : 18.0.4
   @nx/js             : 18.0.4
   @nx/jest           : 18.0.4
   @nx/linter         : 18.0.4
   @nx/eslint         : 18.0.4
   @nx/workspace      : 18.0.4
   @nx/cypress        : 18.0.4
   @nx/devkit         : 18.0.4
   @nx/eslint-plugin  : 18.0.4
   @nx/react          : 18.0.4
   @nx/storybook      : 18.0.4
   @nrwl/tao          : 18.0.4
   @nx/vite           : 18.0.4
   @nx/web            : 18.0.4
   typescript         : 5.1.6
   ---------------------------------------
   Community plugins:
   @koliveira15/nx-sonarqube : 3.5.1
   @nx-tools/nx-container    : 5.2.0
   @nxrocks/nx-spring-boot   : 9.2.2

### Expected Behaviour

Cached projects should be restored to both "target" folder and the ~/.m2/repository folder

### Actual Behaviour

when building the project for the first time or with `--skip-nx-cache` it all works fine, `mvn install` works, and both the target and the ~/.m2/repository folder has the artifacts.

but when the project is restored from the cache, only the "target" folder is restored making other project that depend on this one to not find the jar and fail

### Steps to reproduce the behaviour

1. `nx run back-ends-infra-infra.core:install`
2. `rm -rf ~/.m2/repository/...`
3. `rm -rf .target`
4. `nx run back-ends-infra-infra.core:install`

**Comment by aramfe** [comment]
Agreed. This problems makes the plugin more or less unusable for NX DTE, sadly. Would appreciate help here.

**Comment by tinesoft** [comment]
Hi @SchwarzAmihay 

Thanks for using the plugin and  for reporting.

> Cached projects should be restored to both "target" folder and the ~/.m2/repository folder

I disagree,  the plugin does indeed cache the `target` folder, which make sense because that folder is inside your project, and was generated when running the `build` command. 
Caching the  artifact in the  **external** `~/.m2/repository/...`  folder would make no sense, as this artifact is already present in the `target` folder (where it is first created by Maven, before being pushed to Maven Local repository, on `install`)

The actual issue here, is that the `install`  target was not re-rerun when the dependent  project was restored, meaning that the artifact was not published to Maven Local repo (`~/.m2/repository/...`)


I'm currently working to fix that issue.


Stay tuned!

**Comment by SchwarzAmihay** [comment]
Thank you @tinesoft 
you are doing an amazing job with this plugin!

In the meantime, I will share a workaround to the problem.
I created a new target named `install-to-target` that will build the project or bring it from the cache, and changed the `install` to only install the target to the `~/.m2/repository/...`

```
"targets": {
    "install": {
      "cache": false,
      "executor": "nx:run-commands",
      "options": {
        "cwd": "{projectRoot}",
        "command": "mvn jar:jar install:install"
      },
      "dependsOn": [
        "^install",
        "install-to-target"
      ]
    },
    "install-to-target": {
      "cache": true,
      "executor": "@nxrocks/nx-spring-boot:install",
      "options": {
        "root": "{projectRoot}",
        "args": [
          "-Dmaven.test.skip.exec=true"
        ]
      },
      "dependsOn": [
        "^install-to-target"
      ],
      "outputs": [
        "{projectRoot}/target"
      ]
    },
    ```

**Comment by tinesoft** [comment]
Thanks for the kind words @SchwarzAmihay :-)
Don't hesitate to star ⭐  the project on Github, it helps ^^ 


Thanks for the suggested workaround. I will try to integrate it in plugin directly.

The main takeaway here for me, is **to no longer "cache" the "install" target**, because otherwise, Maven/Gradle will not be invoked to perform the publish to Maven local repository, because Nx will not call it, if found in its cache...

**Comment by aramfe** [comment]
If the "install" targets are not cached, does this essentially make caching useless, or am I missing something?

I believe the main takeaway is to ensure that, even though the install target is cached, it is nonetheless copied into the local Maven repository.

This is especially necessary in the context of distributed task executions, where there is no shared `.m2` directory for all the install targets, especially in the context of multi-module projects, where you want to enable distributed task execution, e.g., executing the packages' install target respectively on different machines, to maximize speed.

This way, a dependent project B package will still resolve its dependency on project A correctly by essentially applying the target into the` .m2` directory.

**Comment by SchwarzAmihay** [comment]
@tinesoft , 
I am not sure if it's relevant to this or another bug but I just noticed that the calculated cache hash does not take into consideration the OS, As I see in my Jenkins CI pipeline reference to my local development environment:
![image](https://github.com/tinesoft/nxrocks/assets/131960634/fe6e8482-f05b-4970-899d-b2996e41036a)

**Comment by aramfe** [comment]
Hey @tinesoft,

thanks for your lengthy response!

I'm not sure if I would use something like caching in GitHub Actions, as there always possible side effects. I'm also not a big fan of caching node_modules for example, as there is always some kind of possible side effect, one wont be able to determine at first sight. 

I still got it running, by simply adjusting the .m2 directory to be in the maven multi module and specifcally also cache the output of the install artifacts in this .m2:

`    "install": {
      "executor": "@nxrocks/nx-spring-boot:install",
      "options": {
        "args": ["-Dmaven.repo.local=.m2"],
        "root": "{workspaceRoot}/backend/packageA",
        "runFromParentModule": true
      },
      "dependsOn": ["^install"],
      "outputs": ["{workspaceRoot}/backend/packageA/target", "{workspaceRoot}/backend/.m2/my/test/package/packageA"],
      "cache": true
    },`

This way it caches both, the .m2 output and the target output. And I've not encountered any issues with that approach so far. This might be helpful for others, who want to use DTE with your plugin (which is amazing by the way!).

**Comment by aramfe** [comment]
@SchwarzAmihay 

You can use the the "inputs" field in your target like this (if python is installed for example): 

`  "inputs": [{ "runtime": "python -c 'import platform; platform.system()'" }],`

This way the hash also takes the output of this command/the current platform into it's hash calculation.

**Comment by jbadeau** [comment]
You can provide maven m2 caching by using Nx's ability to dynamically create target options. The [jnxplus](https://github.com/khalilou88/jnxplus/blob/main/packages/nx-maven/src/graph/create-nodes.ts) maven plugin already does this.
